### ISSUE TITLE
Have the possibility to have a label in addition to the shape on an overlay

### ISSUE THREAD

**Issue by csouchet** [maxGraph/maxGraph#29]
If we have the possibility to change the shape of an overlay, like suggetsted on #28, it would be great to have a label on an overlay.

To do that we should be inspired by what it's done for the Vertex & the Edge (`MxCell`). 
Each `MxCell` has a state defined on [`mxGraphView`](https://github.com/jgraph/mxgraph/blob/master/javascript/src/js/view/mxGraphView.js#L168).
Each state has as property:
- `shape`: The shape of the `MxCell` -> a subclass of mxShape
- `text`: The shape of the label of the `MxCell` -> mxText
- `overlays`: The shapes of all the overlays of the `MxCell` -> An array of mxImageShape (defined on [mxCellRenderer.prototype.createCellOverlays](https://github.com/jgraph/mxgraph/blob/master/javascript/src/js/view/mxCellRenderer.js#L559))

So, to have a label on an overlay, we need to modify the property `overlays` of the state to store 2 shapes by overlay, and modify all the methods where it's used.

**Comment by mcyph** [comment]
Can you paste a simple drawing or diagram of what you'd like to do?

The immediate questions I can see from this are things like where the text will be positioned relative to the overlay (locations within/left/right/top/bottom) and whether the positioning will be automatic or fixed.

**Comment by csouchet** [comment]
Something like this, but with another shape for the overlay:
![image](https://user-images.githubusercontent.com/4921914/116859870-f8423c80-ac00-11eb-895e-1e491a6671c7.png)

**Comment by mcyph** [comment]
Does the example at https://jgraph.github.io/mxgraph/javascript/examples/secondlabel.html do what you're thinking of? (right click and go "view source" to see the code)

Currently the examples and core library are being rewritten without an exact time frame for release, but if there's something more specific we might be able to add it when things become closer to being production-ready.

**Comment by csouchet** [comment]
The second label is not exactly what I need.

For my use case, I need to add one or several overlay on a Cell (Vertex/Edge) on different positions and with different styles (color, shape...) and with different handlers (also different from the handler of the Cell).

Currently I customized, in my code, MxGraph to have different shapes and different position for the overlay, but to have a label in the same time, it's to complicated without rewriting much of the MxGraph library.

**Comment by mcyph** [comment]
Yeah, I'll see if I can get it working with the current release version on the weekend. As you say a rewrite of significant parts may be needed to make this feasible though - a major part of the reason we're refactoring the base code and examples at the moment is to more easily allow for this kind of functionality without making global changes.

**Comment by mcyph** [comment]
You could assign different cell rendering functionality by referring by the ID passed to `mxGraph.insertVertex`: See also https://gist.github.com/mcyph/ecb891574abc6387c3d2dcd8b60df87e for an example of how this could be achieved.

Looking at the example, it is also possible to do this to some degree by creating child vertices then setting the geometry offset:

```javascript
// Gets the default parent for inserting new cells. This
// is normally the first child of the root (ie. layer 0).
var parent = graph.getDefaultParent();

// Adds cells to the model in a single step
graph.getModel().beginUpdate();
try {
    var v1 = graph.insertVertex(parent, null, 'Hello,', 30, 40, 80, 30);
    // Alternative solution of creating a second label by creating a relative child vertex
    // with size (0, 0). This will not be selectable and only the label colors can be used
    // for coloring as the actual shape will have zero size.
    var v11 = graph.insertVertex(v1, null, 'World', 1, 1, 0, 0, 'align=left;verticalAlign=top;labelBackgroundColor=red;labelBorderColor=black', true);
    v11.geometry.offset = new mxPoint(-8, -8);

    var v2 = graph.insertVertex(parent, null, 'World!', 200, 150, 80, 30);
    // Another alternative solution of creating a second label as a relative child vertex
    // but this time with an automatic size so that the cell is actually selectable and
    // the background is painted as a shape.
    var v21 = graph.insertVertex(v2, null, 'World', 1, 1, 0, 0, 'align=left;verticalAlign=top;fillColor=red;rounded=1;spacingLeft=4;spacingRight=4', true);
    v21.geometry.offset = new mxPoint(-8, -8);

    graph.updateCellSize(v21);
    var e1 = graph.insertEdge(parent, null, '', v1, v2);
} finally {
    // Updates the display
    graph.getModel().endUpdate();
}
```

Ideally I think the shapes themselves would be created in an object oriented way and instantiated more directly, e.g. `graph.insertVertex(parent, id, new RectangleShape(...))` but will need to discuss this with other contributors prior to next release as it would need signficant changes.

**Comment by csouchet** [comment]
Thank you for your proposal 🙂 

I have already studied this trail. Unfortunately, this is not completely what I want to do 🙁 
I want to have specific handlers for each overlay, and I want to be able to remove them easily without having to filter out all the child cells of a vertex.

@tbouffard @aibcmars I don't know if I miss something.

**Comment by tbouffard** [comment]
Thanks @mcyph for sharing the gist, the idea of using the relative coordinates and offset is nice to manage the overlay position.

As @csouchet mentioned, the gist doesn't provide everything that the current mxCellOverlays support.
Overlays are separated from the regular cells in the mxgraph model (and are rendered in a dedicated SVG group), so we can identify them easily and remove them (all or one) from the cell they belong too with mxgraph API. Managing them as cell children also require some extra processing. This is not something complicated but this is less integrated.

Anyway, we have needs in [bpmn-visualization](https://github.com/process-analytics/bpmn-visualization-js) that we are currently refining and we will also conduct extra investigation that we will be able to share here in the future.
